src/quic/session.cc

```diff
@@ -0,0 +1,3506 @@
+    const SessionTicketAppData& app_data, SessionTicketAppData::Flag flag) {
+  DEBUG(session_, "Application getting session ticket app data");
+  return flag == SessionTicketAppData::Flag::STATUS_RENEW
+             ? SessionTicketAppData::Status::TICKET_USE_RENEW
+             : SessionTicketAppData::Status::TICKET_USE;
+}
+
+BaseObjectPtr<Packet> Session::Application::CreateStreamDataPacket() {
+  return Packet::Create(
+      env(),
+      session_->endpoint_.get(),
+      session_->remote_address_,
+      "stream data",
+      ngtcp2_conn_get_max_udp_payload_size(session_->connection()));
+}
+
+void Session::Application::StreamClose(Stream* stream, Maybe<QuicError> error) {
+  DEBUG_ARGS(session_, "Application closing stream %" PRIi64, stream->id());
+  stream->Destroy(error);
+}
+
+void Session::Application::StreamReset(Stream* stream,
+                                       uint64_t final_size,
+                                       QuicError error) {
+  DEBUG_ARGS(session_,
+             "Application resetting stream %" PRIi64 " with final size %" PRIu64
+             " [%s]",
+             stream->id(),
+             final_size,
+             error);
+  stream->ReceiveResetStream(final_size, error);
+}
+
+void Session::Application::StreamStopSending(Stream* stream, QuicError error) {
+  DEBUG_ARGS(session_,
+             "Application stop sending stream %" PRIi64 " [%s]",
+             stream->id(),
+             error);
+  stream->ReceiveStopSending(error);
+}
+
+void Session::Application::SendPendingData() {
+  PathStorage path;
+
+  DEBUG(session_, "Starting pending data loop");
+  BaseObjectPtr<Packet> packet;
+  uint8_t* pos = nullptr;
+  int err = 0;
+
+  size_t maxPacketCount =
+      std::min(static_cast<size_t>(64000),
+               ngtcp2_conn_get_send_quantum(session_->connection()));
+  size_t packetSendCount = 0;
+  DEBUG_ARGS(session_,
+             "Maximum number of packets to send this iteration: %" PRIu64,
+             maxPacketCount);
+
+  const auto updateTimer = [&] {
+    ngtcp2_conn_update_pkt_tx_time(session_->connection(), uv_hrtime());
+    session_->UpdateTimer();
+  };
+
+  const auto congestionLimited = [&](auto packet) {
+    auto len = pos - ngtcp2_vec(*packet).base;
+    // We are either congestion limited or done.
+    if (len) {
+      // Some data was serialized into the packet. We need to send it.
+      packet->Truncate(len);
+      session_->Send(std::move(packet), path);
+    }
+
+    updateTimer();
+  };
+
+  for (;;) {
+    ssize_t ndatalen;
+    StreamData stream_data;
+
+    DEBUG(session_, "Selecting stream data to encode");
+    err = GetStreamData(&stream_data);
+
+    if (err < 0) {
+      session_->last_error_ = QuicError::ForNgtcp2Error(NGTCP2_ERR_INTERNAL);
+      return session_->CloseSilently();
+    }
+
+    DEBUG_ARGS(session_, "Stream Data: %s", stream_data);
+
+    if (!packet) {
+      packet = CreateStreamDataPacket();
```

Comment:
This packet needs to retain the outbound source or node may crash

See https://github.com/HalleyAssist/node/commit/60d9bbcc7174ac2a3bc1ba8de442afe7aa938e7a for how I did it